src/ColumnsPaginatedBookView.ts

```diff
@@ -83,7 +83,21 @@ export default class ColumnsPaginatedBookView implements PaginatedBookView {
         const images = body.querySelectorAll("img");
         for (const image of images) {
             image.style.maxWidth = width;
-            image.style.maxHeight = this.height + "px";
+
+            // Determine how much vertical space there is for the image.
+            let nextElement = image;
+            let totalMargins = 0;
+            while (nextElement !== body) {
+                const computedStyle = window.getComputedStyle(nextElement);
+                if (computedStyle.marginTop) {
+                    totalMargins += parseInt(computedStyle.marginTop.slice(0, -2), 10)
+                }
+                if (computedStyle.marginBottom) {
+                    totalMargins += parseInt(computedStyle.marginBottom.slice(0, -2), 10)
+                }
+                nextElement = nextElement.parentElement;
+            }
+            image.style.maxHeight = (this.height - totalMargins) + "px";
```

Comment:
This is better than before, but it still doesn't fix #122 in all cases. For example. the publisher might put an empty <p> tag in front of the image instead of making it a parent. But it's hard to distinguish that from actual content.